### ISSUE TITLE
thunderbird: large emails break firenvim?

### ISSUE THREAD

**Issue by glacambre** [glacambre/firenvim#1263]
xxxx xx.x xx xxx xxxxxxxxx xx xxxxxx xx.xx xxx xx /xxx/xxxxx/xxxx.
xx xxxxxxx x xxx xxxxxxxx xxx xxx_xxxxxx xxxxxxxxx xxx xxxxxxx x xxx
xxxxxxxx.
xxxx xx xx xxxxxxx xx x xxxx xxxxx xxxxx xxxxxxxxx xx xxx xx xxxxxxxx xxxx
xx xxxxxx xx:

@xxx_xxxxxx(
      xxx=xx<xxxxxx_x_x_xx_xx>,
      xxxx_xxxx=xx<xxxxxx_xxxx_xxxx>,
      xxxxxxxx=xx"xxx_xx-xxxxx",
      xxxxxxx=xx"xxxxxxxx.xx",
      xxx=xxxxxx,
)

xxxx xxxx xxx xxxxxx xx xxxx xxxx xxxxxxxxx xxx xxx xxxxxxxxx:

x xxxx: xx xxx
x xxxx: xx xxx
xx xxxx: xx xxx

xx xxx, xxx x, xxxx xx x:xx xx xxxxxxx xxxxx <xxxxx@xxxxxxx.xxx> xxxxx:

> xx xxxxxxxxx xx xxxxx xxx'x xxxx xxxx xx xx xxxxxxxx xx xxx xxxxxxxxx (xx
> xxxxxx xxxxxx xxxxxx xxxx xxx xxxxxxxxxx xxxxx xxxxxx ;-) )
>
> xx xx xxxxxxxxx xx xxxxx xxxx'x xx xx xxxxxx xx xxx xxxx xxxx.xxxxxxx
> xxxxxxxxx
>
> xx xxxxxxxx xxxx x xxx xx xxxxxxxxxx xxxxxx xxxxxx xxxxx xxxx xxxxxxx xxx=
xx
>
> --
> xxxxxxx
>
> xx xxx, xxx x, xxxx xx x:xx xx xxxxxx xxxxxxxxx <xxxxxxxxx@xxxxxxx.xxx>
> xxxxx:
>
>>
>> x xxxxx xxxxx xx xx xxxx xxx x xxx xx xxxxxxx xxxx
>>
>> x xxxxxxx xxxx x xxx xxxx xx xxxxxxx xxxx xx xxx xxxxxx xx.xx xxx xxxxxx=
x
>> xxx xxx.
>>
>> x xxxxxx xxx xxxxx xxxx x xxxx.xxxxxxx xxxxxxxx xxxx xxxxxxxxxxx xxx
>> xxxxxxxxxxxxx xxxxxxx xxxx xxx xxxxxxxx xxxx xxxx.
>> xxxx xxxxx xxx xx xxx xxx xx xxxxxxxx. xxx xxxxxxx xxxx xx xxx xxxxxx xx
>> xxxxxxxxx xxxx:
>>
>> xxxxxxx xx...
>>
>>
>>  xxxxxx x: xxxxx xx
>>  xxxxx xxxx
>>
>>
>>  xxxx xx xxxxxx xxxxxxxxx, x xxxxxx(x)
>>  xxxxxx xxxxxx xxxxxxxxx
>>    xxxxxxxx xxxx (xxxxx) xxxxxxxxx(xx/x)
>>    xxxxxxxx x.x
>>
>>
>>  xxxxxx xx xxxx xxxxxxxxx, x xxxxxx(x)
>>  xxxxxx xxxxxx xxxxxxxxx
>>    xxxxxxxx xxxx (xxxxx) xxxxxxxxx(xx/x)
>>    xxxxxxxx x.x
>>
>>
>>  xxxxxx xx xxxxxx xxxxxxxxx, x xxxxxx(x)
>>  xxxxxx xxxxxx xxxxxxxxx
>>    xxxxxxxx xxxx (xxxxx) xxxxxxxxx(xx/x)
>>    xxxxxxxx xxx.x
>>
>>
>> xxx xxxx xxxxx xxxx xx x xxx xxxxxxxx xx xx xx xxxx xxxx x xxxxx xx xxx =
x
>> xxxx.xxxxxx x xxx xx xxxxxxxxxxxxxxxxxxxxxxxxxxxx
>> xxxxx xxxx xxx. xxxxx xx xxxxx xxx xxxx xxxxxx xxx xxxx xxxxxxxx xxxx xx=
x
>> x xxx'x xxxx xxx xx xxxxxx xxxxxxxx xxxx
>> xxxxxxxxxxxxxx xx xxxxxx. @xxxxxxx xxxxx <xxxxxxx.xxxxx@xxxxxxx.xxx> xxx=
x
>> xx xxxxxxx xxx xxxx xxxx xx xxxxxx xxxx xxxxxxx xxxxx?
>>
>>
>>
>> xx xxx, xxx x, xxxx xx x:xx xx xxxxxx xxxxxxxxx <xxxxxxxxx@xxxxxxx.xxx>
>> xxxxx:
>>
>>> xxxxxx x xxx xxxx xxx xxxxxx.
>>>
>>>
>>> > xx x xxx. xxxx =xx=xx xx:xx, xxxxxxx xxxxxxxx <xxxxxxxx@xxxxxxx.xxx> =
x
>>> =xx=xxxxxx :
>>> >
>>> > xx xx/x/xx xx:xx, xxxxxxx xxxxx xxxxx:
>>> > > xxx xxx xxxxxxx x xxxxx xxx xxx xxxx xxx xxxxxx.
>>> >
>>> > xxxxxx :).
>>>
>>>

**Comment by glacambre** [comment]
Turns out this is https://bugzilla.mozilla.org/show_bug.cgi?id=1746468 .

**Comment by doronbehar** [comment]
Is firenvim broken with thunderbird 92? I just tried to install it and I don't know what's supposed to happen, firenvim works in my browser. Should I suppose to be able to compose messages with firenvim?

**Comment by glacambre** [comment]
Yes it is broken. You can't expect Firenvim to work with any Thunderbird version for now and I don't think you can expect it to work in the future either, as Thunderbird maintainers do not seem to care much about the issues I report or the patches I send them (which is okay, they don't owe me anything).

**Comment by cycomanic** [comment]
@glacambre  I was looking at getting firenvim working on thunderbird and came across this post. 

I was trying to see if I could maybe add some comments on the thunderbird issue to maybe get the team there to care more about the problems and saw that they are asking for clarification. Maybe you could add the information? 

Or alternatively I'm happy to try and bisect where the problem came from, but I'm not familiar with mozilla extension development, how would I run the reproducer that you attached in the issue?

**Comment by paride** [comment]
Still not working with Thunderbird 102 :(

**Comment by glacambre** [comment]
@cycomanic In order to run the reproducer, you need to download it and unpack it.

Then, in Thunderbird, click the "hamburger menu" (the three horizontal bars) at the top right corner of the screen. Select "Addons", this should open the Add-ons Manager. Once there, click on the cog icon. Select "Debug Add-On", then "Load Temporary Add-On" and choose `manifest.json`.

Once the reproducer is loaded, open a compose window (e.g. by clicking on the button to write a new mail). Working thunderbird versions will display a black square, broken thunderbird versions won't.

**Comment by glacambre** [comment]
@cycomanic I saw you did the work of bisecting the problem on bugzilla. Thanks a lot, I'm not sure you'll be able to get Mozilla to do anything about the problem but I appreciate the time you spent of this.

**Comment by cycomanic** [comment]
@glacambre it seems thunderbird disabled javascript in compose windows in response to CVE-2021-43528 which will prevent the canvas element from displaying. I suspect that they will not just reenable it. There is the composeScripts API could using that be a solution for firenvim (I don't have much experience with JS and firefox extensions, so I might be completely off).

**Comment by glacambre** [comment]
On 7/25/22 22:23, Jochen Schröder wrote:
> @glacambre <https://github.com/glacambre> it seems thunderbird disabled 
> javascript in compose windows in response to CVE-2021-43528 
> <https://github.com/advisories/GHSA-q4gq-74cq-85g9> which will prevent 
> the canvas element from displaying. I suspect that they will not just 
> reenable it. There is the composeScripts API could using that be a 
> solution for firenvim (I don't have much experience with JS and firefox 
> extensions, so I might be completely off).

Unfortunately the composeScript API is exactly what the reproducer uses:

$tar -xvf repro.tar
thunderbird_repro/
thunderbird_repro/background.js
thunderbird_repro/manifest.json
thunderbird_repro/compose.js
$cd thunderbird_repro
$cat background.js
browser.composeScripts.register({
     js: [{ file: "compose.js" }],
});
$cat compose.js

try {
     document.body.innerText += "compose script loaded";
     const canvas = document.createElement("canvas");
     document.body.appendChild(canvas);
     const context = canvas.getContext("2d");
     context.fillStyle = "black";
     context.fillRect(10, 10, 100, 100);
     // const txt = document.body.innerHTML;
     // document.body.innerText += txt;
} catch (e) {
     document.body.innerText += "failure: " + e.toString();
}


I don't quite understand how disabling JS in the compose window can 
prevent the canvas from working while leaving DOM manipulation fully 
functional (we know DOM manipulation works because we get the "compose 
script loaded" message in the compose window). Very puzzling...

**Comment by cycomanic** [comment]
>> On 7/25/22 22:23, Jochen Schröder wrote:
>> [@glacambre](https://github.com/glacambre) <https://github.com/glacambre> it seems thunderbird disabled
> javascript in compose windows in response to CVE-2021-43528
>><https://github.com/advisories/GHSA-q4gq-74cq-85g9> which will prevent
>>the canvas element from displaying. I suspect that they will not just
>> eenable it. There is the composeScripts API could using that be a
>> solution for firenvim (I don't have much experience with JS and firefox
>> extensions, so I might be completely off).

> Unfortunately the composeScript API is exactly what the reproducer uses:

Yes I just realised that after reading some more as well. 

>I don't quite understand how disabling JS in the compose window can
prevent the canvas from working while leaving DOM manipulation fully
functional (we know DOM manipulation works because we get the "compose
script loaded" message in the compose window). Very puzzling...

The way I understand it after a bit of reading and with my (very limited) understanding of JS is, that the canvas element requires JS to render, that is not happening, so the canvas element is created by executing compose.js via the composeScripts api, but the canvas is not rendered. If I change compose.js to:
```javascript

try {
    document.body.innerText += "compose script loaded2";
    const canvas = document.createElement("canvas");
    //document.body.appendChild(canvas);
    const context = canvas.getContext("2d");
    context.fillStyle = "black";
    context.fillRect(10, 10, 100, 100);
    // const txt = document.body.innerHTML;
    // document.body.innerText += txt;
    const imgUrl = canvas.toDataURL();
    const img = document.createElement('img');
    img.src = imgUrl;
    document.body.appendChild(img)

} catch (e) {
    document.body.innerText += "failure: " + e.toString();
}
```
I can see the black element. I'm not sure if that helps for firenvim though?

**Comment by glacambre** [comment]
>I'm not sure if that helps for firenvim though?

No, unfortunately this doesn't help as `canvas.toDataURL()` and `img.src = imgUrl` are far too slow for interactive use. The solution probably is to run the firenvim script in an iframe that has a webextension context, but this doesn't work either due to https://bugzilla.mozilla.org/show_bug.cgi?id=1707449 ...

**Comment by paride** [comment]
Checking how the situation is after two years. Is firenvim still broken/unsupported by the latest Thunderbird versions (128 ESR, 133 beta)?

@glacambre to unblock this, do you have a pointer to a "first blocking bug" on the Thunderbird side that, once fixed, could help getting it to work again?

Thanks!

**Comment by glacambre** [comment]
>Is firenvim still broken/unsupported by the latest Thunderbird versions (128 ESR, 133 beta)?

Yes.

>do you have a pointer to a "first blocking bug" on the Thunderbird side that, once fixed, could help getting it to work again?

Solving either https://bugzilla.mozilla.org/show_bug.cgi?id=1707449 or https://bugzilla.mozilla.org/show_bug.cgi?id=1746468 would unblock the situation, but I wouldn't spend time on this if I were you, it seems these restrictions were introduced on purpose by the Thunderbird people.

**Comment by paride** [comment]
Thanks. I'll take your suggestion; too bad this is not available on Thunderbird, having a good editor there would really make a difference.

Unrelated: nice clown avatar by Edward Hopper. :)

**Comment by stevenxxiu** [comment]
You could use <https://github.com/Frederick888/external-editor-revived/> if you want a custom editor to compose email. I configured my *Thunderbird* to use *Neovim* and edit in *Markdown*. See <https://github.com/stevenxxiu/external_editor_revived_markdown>.
